x-pack/plugins/aiops/server/routes/explain_log_rate_spikes.ts

```diff
@@ -211,31 +216,148 @@ export const defineExplainLogRateSpikesRoute = (
         }
 
         if (groupingEnabled) {
+          // To optimize the `frequent_items` query, we identify duplicate change points by count attributes.
+          // Note this is a compromise and not 100% accurate because there could be change points that
+          // have the exact same counts but still don't co-occur.
+          const duplicateIdentifier: Array<keyof ChangePoint> = [
+            'doc_count',
+            'bg_count',
+            'total_doc_count',
+            'total_bg_count',
+          ];
+
+          // These are the deduplicated change points we pass to the `frequent_items` aggregation.
+          const deduplicatedChangePoints = dropDuplicates(changePoints, duplicateIdentifier);
+
+          // We use the grouped change points to later repopulate
+          // the `frequent_items` result with the missing duplicates.
+          const groupedChangePoints = groupDuplicates(changePoints, duplicateIdentifier).filter(
+            (g) => g.group.length > 1
+          );
+
           const { fields, df } = await fetchFrequentItems(
             client,
             request.body.index,
             JSON.parse(request.body.searchQuery) as estypes.QueryDslQueryContainer,
-            changePoints,
+            deduplicatedChangePoints,
             request.body.timeFieldName,
             request.body.deviationMin,
             request.body.deviationMax
           );
 
-          // Filter itemsets by significant change point field value pairs
-          const filteredDf = df.filter((fi) => {
-            const { set: currentItems } = fi;
+          // The way the `frequent_items` aggregations works could return item sets that include
+          // field/value pairs that are not part of the original list of significant change points.
+          // This cleans up groups and removes those unrelated field/value pairs.
+          const filteredDf = df
+            .map((fi) => {
+              const { set: currentItems } = fi;
+
+              for (const [field, value] of Object.entries(currentItems)) {
+                const stIndex = changePoints.findIndex(
+                  (cp) => cp.fieldName === field && cp.fieldValue === value
+                );
+                if (stIndex === -1) {
+                  delete currentItems[field];
```

Comment:
Deleting items while in the middle of looping over them can be dangerous. I appreciate there is no counter or index being used here, but perhaps adding them into a new set would be cleaner?
Something like:
```ts
fi.set = {};
Object.entries(currentItems).forEach(([field, value]) => {
  if (
    changePoints.find((cp) => cp.fieldName === field && cp.fieldValue === value) !==
    undefined
  ) {
    fi.set[field] = value;
  }
});
fi.size = Object.keys(fi.set).length;
```

But feel free to ignore if you think this is risk free.

Reply:
Good point! I refactored the code to work the other way around, it nows builds a new set based on the field/values that match. Update in [684ef1c](https://github.com/elastic/kibana/pull/140891/commits/684ef1c9214c50ff77c59bc20e569ef00c0fb275).